baseapp/abci.go

```diff
@@ -266,7 +278,19 @@ func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) abci.Respon
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) ProcessProposal(req abci.RequestProcessProposal) abci.ResponseProcessProposal {
-	// TODO: Implement.
+	ctx := app.checkState.ctx
+
+	for _, txBytes := range req.Txs {
+		anteCtx, _ := app.cacheTxContext(ctx, txBytes)
+		tx, err := app.txDecoder(txBytes)
+		if err != nil {
+			return abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}
+		}
+		ctx, err = app.anteHandler(anteCtx, tx, false)
```

Comment:
Why are you not running this on `PrepareProposal`?

Reply:
@JeancarloBarrios this doesn't look quite correct. In the ADR, we allude to the app having full control over `ProcessProposal` for better customizability. This means, we shouldn't be calling  `anteHandler` directly. We should be calling `app.processProposal`, where the default implementation executes the entire proposal.

@sergio-mena for the record, we call `AnteHandler` on `CheckTx`. If the tx passes, it is inserted into the app-mempool (and Tendermint's mempool too). `PrepareProposal` takes txs from the app-side mempool. I believe the idea we now have is that `ProcessProposal` executes the entire proposal, i.e. `AnteHandler` + `Msg Execution`.

Reply:
Yes I have updated this i still need to push but I updated after our conversation yesterday

Reply:
@alexanderbez be careful here folks. Let me describe an example and you tell me if that can happen.

1. Tendermint calls `CheckTx` on txA, SDK calls the `AnteHandler` on txA which uses the Apps's current state; the `AnteHandler` check passes and txA is added to both mempools
2. There are many transactions and txA has low priority, so there are, say, 10 blocks decided that didn't include txA
3. txA is finally selected in line 253 above (`app.mempool.Select`), `PrepareProposal` does not run the `AnteHandler`, txA makes it to the proposal
4. `ProcessProposal` is run on the proposal that includes txA so it runs the `AnteHandler` on txA. There have been 10 blocks decided, so the state has changed. With the new state, the `AnteHandler` check fails on txA
5. `ProcessProposal` returns reject

Can this happen?
If it can, we may be able to solve it by re-running the `AnteHandler` also on `PrepareProposal`